File: macos/Guardian/Service/MonitorService.swift
import AppKit
import CoreGraphics
import Foundation
import UserNotifications

/// The monitoring loop. Two modes meet here:
///
/// **Focus mode** (the product). A session is running — the user declared a task like "revising
/// integration by parts" — so every `intervalSeconds` we capture one frame of whichever watched app
/// is frontmost and ask the model whether that screen belongs to that task.
///
/// **Content rules** (opt-in, off by default). The original behaviour: an always-on classifier
/// checking screens against written content guidelines. Kept because it works, but it is no longer
/// what the app is for.
///
/// The most important structural change: **when no session is running, nothing is captured at
/// all** — not captured and discarded, never taken. A monitor that only looks during a window you
/// opened yourself is a different thing to live with, and that difference is worth the extra branch.
///
/// Cadence follows the session's interval rather than how fast the model answers. The old loop
/// fired 1.2s after each reply, which is right for content safety (one frame of the wrong thing
/// matters) and wrong here: it would burn hundreds of calls an hour on a question whose answer
/// changes over minutes.
@MainActor
final class MonitorService: ObservableObject {

    static let shared = MonitorService()

    @Published private(set) var isRunning = false
    /// Short status line for the dashboard and the menu bar.
    @Published private(set) var status = "Idle"

    private var task: Task<Void, Never>?
    private let prefs = Prefs.shared
    private let log = EventLog.shared
    private let sessions = SessionStore.shared

    /// Per-bundle throttle so persistently-unverifiable apps don't spam the alert.
    private var lastAlertAt: [String: Date] = [:]
    private var screenWasVisible = true
    /// When the next focus check is due, per app. Keyed by app so switching apps checks the new one
    /// promptly instead of inheriting the previous app's countdown — the moment you switch into a
    /// distraction is exactly the moment worth looking.
    private var nextCheckAt: [String: Date] = [:]

    // Content-rules cadence (unchanged from the original app).
    private let minGap: TimeInterval = 1.2
    private let idlePoll: TimeInterval = 2.5
    /// How often we re-check *which* app is frontmost while not on a watched app. Cheap — one
    /// AppKit call, no capture, no inference — so it can be brisk.
    private let focusIdlePoll: TimeInterval = 3.0
    /// After a transient backend failure, retry sooner than a full interval but not instantly.
    private let transientRetry: TimeInterval = 20
    private let alertThrottle: TimeInterval = 10 * 60

    private init() {}

    // MARK: - Lifecycle

    func start() {
        guard task == nil else { return }
        isRunning = true
        prefs.monitoringEnabled = true
        status = "Monitoring active"
        let mode = sessions.isActive ? "focus" : "content-rules"
        log.add("▶︎ monitoring started (\(mode))\(prefs.dryRun ? " [TEST]" : "")")
        nextCheckAt.removeAll()
        task = Task { await loop() }
    }

    func stop() {
        task?.cancel()
        task = nil
        isRunning = false
        prefs.monitoringEnabled = false
        status = "Stopped"
        log.add("■ monitoring stopped")
    }

    private func loop() async {
        while !Task.isCancelled {
            let wait = await tick()
            try? await Task.sleep(nanoseconds: UInt64(max(wait, 0.5) * 1_000_000_000))
        }
    }

    // MARK: - One tick

    /// Do at most one unit of work. Returns how many seconds to wait before the next tick.
    private func tick() async -> TimeInterval {
        // A block overlay is up — we already cover the screen; don't capture or evaluate.
        if BlockController.shared.isBlocking { return idlePoll }

        // Display asleep / locked / screen saver: nobody is looking at anything, and capturing
        // would hand us a black frame that reads as "can't see" and alerts. Pause instead.
        guard ScreenState.isVisible else {
            if screenWasVisible {
                screenWasVisible = false
                log.add("💤 screen \(ScreenState.reason) — monitoring paused")
                status = "Screen \(ScreenState.reason)"
            }
            return idlePoll
        }
        if !screenWasVisible {
            screenWasVisible = true
            log.add("👁️ screen back on — monitoring resumed")
            status = "Monitoring active"
        }

        if let session = sessions.current {
            return await focusTick(session)
        }
        if prefs.contentRulesEnabled {
            return await contentTick() ? minGap : idlePoll
        }

        // Nothing to do: no session, content rules off. Explicitly *not* capturing anything.
        status = "No focus session — not watching"
        return idlePoll
    }

    // MARK: - Focus mode

    private func focusTick(_ session: FocusSession) async -> TimeInterval {
        let interval = TimeInterval(session.intervalSeconds)

        if session.isPaused, let until = session.pausedUntil {
            let left = max(until.timeIntervalSinceNow, 0)
            status = "Paused \(Int(left))s"
            return min(left + 0.5, interval)
        }

        let watchlist = session.watchlist(defaults: prefs.monitoredApps)
        guard let fg = FrontmostApp.bundleId, !fg.isEmpty, watchlist.contains(fg) else {
            // Not a watched app. In focus mode this is the normal, uninteresting case — you are in
            // your editor, or your terminal, or anything you never asked to be policed.
            status = idleStatus(session)
            return focusIdlePoll
        }

        if Overrides.isActive(fg) { return focusIdlePoll }

        // Rate-limit per app so switching back and forth can't force a check storm.
        if let due = nextCheckAt[fg], due > Date() {
            status = idleStatus(session)
            return min(due.timeIntervalSinceNow, focusIdlePoll)
        }

        let name = FrontmostApp.shortName(for: fg)
        let title = FrontmostApp.windowTitle(for: fg)
        let dry = prefs.dryRun

        guard let frame = await ScreenCapturer.capture() else {
            await handleUnverifiable(fg, name, "screen could not be captured (grant Screen Recording?)", dry)
            nextCheckAt[fg] = Date().addingTimeInterval(interval)
            return interval
        }
        if FrameQuality.isUnreadable(frame) {
            await handleUnverifiable(fg, name, "screen is blank or capture-protected", dry)
            nextCheckAt[fg] = Date().addingTimeInterval(interval)
            return interval
        }

        status = "Checking \(name)…"
        let cfg = prefs.providerConfig()
        let notes = extraNotes(for: session, app: fg)
        let started = Date()
        let verdict = await Providers.evaluate(cfg, image: frame, task: session.task,
                                               appName: name, windowTitle: title,
                                               extraNotes: notes,
                                               onKeyWorked: { key in
            Task { @MainActor in Prefs.shared.promoteApiKey(key) }
        })
        let ms = Int(Date().timeIntervalSince(started) * 1000)

        // Transient backend trouble is never evidence about the user: log, retry sooner, no block.
        if verdict.transient {
            log.add("⏳ \(name) — AI unavailable (\(verdict.reason)) — skipped")
            status = "AI unavailable — retrying"
            nextCheckAt[fg] = Date().addingTimeInterval(transientRetry)
            return transientRetry
        }

        nextCheckAt[fg] = Date().addingTimeInterval(interval)

        if verdict.undetermined {
            await handleUnverifiable(fg, name, "AI could not analyse (\(verdict.reason))", dry)
            return interval
        }

        if verdict.onTask {
            sessions.recordCheck(offTask: false)
            Judgements.record(session: session, app: fg, appName: name, windowTitle: title,
                              verdict: "on_task", reason: verdict.reason,
                              confidence: verdict.confidence, action: "allowed",
                              provider: cfg.describe)
            log.add("✅ \(name) on task (\(ms)ms, \(cfg.model))")
            status = idleStatus(session)
            return interval
        }

        // Off task.
        sessions.recordCheck(offTask: true)
        let action = dry ? "logged" : "blocked"
        let jid = Judgements.record(session: session, app: fg, appName: name, windowTitle: title,
                                    verdict: "off_task", reason: verdict.reason,
                                    confidence: verdict.confidence, action: action,
                                    provider: cfg.describe)
        if dry {
            log.add("🟠 \(name) WOULD BLOCK — off task (\(ms)ms) — \(verdict.reason) [TEST]")
            status = "WOULD block \(name)"
        } else {
            log.add("🚫 \(name) OFF TASK (\(ms)ms) — \(verdict.reason)")
            await enforceOffTask(session, fg, name, verdict, jid)
        }
        return interval
    }

    private func idleStatus(_ session: FocusSession) -> String {
        guard let remaining = session.remaining else { return "Focus: \(session.task.prefix(40))" }
        return "Focus: \(session.task.prefix(32)) (\(Int(remaining) / 60)m left)"
    }

    /// Standing notes handed to the classifier: the user's own, plus anything the experimental
    /// learner has concluded. Both capped hard — a prompt suffix that grows without bound
    /// eventually costs more than the screenshot does.
    private func extraNotes(for session: FocusSession, app: String) -> String {
        var notes = [prefs.focusNotes.trimmingCharacters(in: .whitespacesAndNewlines)]
        if prefs.learningEnabled {
            notes.append(LearnedPolicy.promptSuffix(task: session.task, app: app))
        }
        return String(notes.filter { !$0.isEmpty }.joined(separator: "\n").prefix(1500))
    }

    /// Raise the block. What the block *offers* depends on the accountability level the session was
    /// started at — read from the session, not from a global setting that could have drifted since.
    private func enforceOffTask(_ session: FocusSession, _ bundleId: String, _ name: String,
                                _ verdict: Providers.Verdict, _ judgementId: String) async {
        prefs.lastViolationAt = Date()
        status = "BLOCKED \(name)"
        BlockController.shared.show(bundleId: bundleId, appName: name, reason: verdict.reason,
                                    session: session, judgementId: judgementId)
        notifyLocal("Off task in \(name)", verdict.reason.isEmpty ? session.task : verdict.reason)

        if session.accountability.alertsPartner {
            await sendAlert(bundleId, "[Lockout] Off task in \(name)",
                            "Task: \(session.task)\n\n\(name) was blocked.\nReason: \(verdict.reason)",
                            throttle: false)
        }
    }

    // MARK: - Content-rules mode (opt-in; the original behaviour)

    /// Returns true if a monitored app was on screen and was actually checked.
    private func contentTick() async -> Bool {
        let monitored = prefs.monitoredApps
        guard let fg = FrontmostApp.bundleId, !fg.isEmpty, monitored.contains(fg) else {
            return false
        }
        if Overrides.isActive(fg) { return false }

        let dry = prefs.dryRun
        let name = FrontmostApp.shortName(for: fg)

        guard let frame = await ScreenCapturer.capture() else {
            await handleUnverifiable(fg, name, "screen could not be captured (grant Screen Recording?)", dry)
            return true
        }
        if FrameQuality.isUnreadable(frame) {
            await handleUnverifiable(fg, name, "screen is blank or capture-protected", dry)
            return true
        }

        let started = Date()
        let cfg = OllamaClient(prefs: prefs).config()
        let verdict = await OllamaClient(prefs: prefs).evaluate(frame, config: cfg)
        let ms = Int(Date().timeIntervalSince(started) * 1000)

        if verdict.violation {
            if dry {
                log.add("🟠 \(name) WOULD BLOCK (\(ms)ms) — \(verdict.reason) [TEST]")
            } else {
                log.add("🚫 \(name) VIOLATION (\(ms)ms) — \(verdict.reason)")
                prefs.lastViolationAt = Date()
                BlockController.shared.show(bundleId: fg, appName: name, reason: verdict.reason)
                notifyLocal("Guardian blocked \(name)", verdict.reason)
                await sendAlert(fg, "[Lockout] Blocked \(name)",
                                "Content rule triggered in \(name).\n\nReason: \(verdict.reason)",
                                throttle: false)
            }
            return true
        }
        if verdict.transient {
            log.add("⏳ \(name) AI busy — skipped (\(ms)ms) — \(verdict.reason)")
            return true
        }
        if verdict.undetermined {
            await handleUnverifiable(fg, name, "AI could not analyse (\(verdict.reason))", dry)
            return true
        }
        log.add("✅ \(name) cleared (\(ms)ms, \(cfg.model))")
        return true
    }

    // MARK: - Shared

    /// A watched app's screen could NOT be read (blank/protected frame, capture failure, or an
    /// answer we couldn't parse). Never blocks. This is the anti-lockout rule: a screen we cannot
    /// see is not a screen we get to punish.
    private func handleUnverifiable(_ bundleId: String, _ name: String, _ why: String,
                                    _ dry: Bool) async {
        if dry {
            log.add("🙈 \(name) CAN'T SEE — \(why) — would alert [TEST]")
            status = "Can't see \(name)"
            return
        }
        log.add("🙈 \(name) CAN'T SEE — \(why)")
        status = "Can't see \(name)"

        let session = sessions.current
        let alerts = prefs.alertOnUnverifiable
            && (session == nil || session!.accountability.alertsPartner)
        if alerts {
            await sendAlert(bundleId, "[Lockout] Couldn't verify \(name)",
                            "Guardian could not see what was on screen in \(name) (\(why)).",
                            throttle: true)
        }
        if prefs.closeUnverifiable { FrontmostApp.hide(bundleId: bundleId) }
    }

    private func sendAlert(_ key: String, _ title: String, _ body: String, throttle: Bool) async {
        if throttle {
            if let last = lastAlertAt[key], Date().timeIntervalSince(last) < alertThrottle { return }
            lastAlertAt[key] = Date()
        }
        let fullBody = "\(body)\n\n\(Date())"
        let cfg = Pusher.Config(enabled: prefs.pushEnabled, server: prefs.ntfyServer,
                                topic: prefs.ntfyTopic)
        let result = await Pusher.send(cfg, title: title, message: fullBody)
        if case .failure(let e) = result {
            log.add("⚠️ push failed: \(e.localizedDescription)")
        }
    }

    private func notifyLocal(_ title: String, _ body: String) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        let req = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
        UNUserNotificationCenter.current().add(req)
    }
}

File: windows/guardian/service/monitor_service.py
"""The monitoring loop.

Focus mode runs while a session is active: every `interval_seconds` it screenshots whichever
watched app is in front and asks the model whether that screen belongs to the declared task.
Content-rules mode is the original always-on classifier, opt-in and off by default.

When no session is running and content rules are off, nothing is captured at all.

Cadence follows the session's interval rather than how fast the model answers. The old loop fired
1.2s after each reply, which is right for content safety and wrong here: it would burn hundreds of
calls an hour on a question whose answer changes over minutes.
"""

import threading
import time
from datetime import datetime

from ..ai import frame_quality, providers
from ..ai.ollama_client import OllamaClient
from ..capture import foreground_app, screen_capturer, screen_state
from ..focus.session import SessionStore
from ..models import judgements
from ..models.event_log import EventLog
from ..models.prefs import Prefs
from ..notify import notify_local
from ..push import pusher
from . import overrides
from .block_controller import BlockController

# Content-rules cadence, unchanged from the original app.
MIN_GAP = 1.2
IDLE_POLL = 2.5

# Focus mode: how often to re-check which app is in front while not on a watched app. Cheap — a
# Win32 call, no capture, no inference.
FOCUS_IDLE_POLL = 3.0
TRANSIENT_RETRY = 20.0

ALERT_THROTTLE = 10 * 60


class MonitorService:

    _instance = None
    _lock = threading.RLock()

    def __init__(self):
        self.is_running = False
        self.status = "Idle"
        self._thread = None
        self._stop = threading.Event()
        self._prefs = Prefs.shared()
        self._log = EventLog.shared()
        self._sessions = SessionStore.shared()
        self._last_alert_at = {}
        self._screen_was_visible = True
        self._unreadable_streak = 0
        # When the next check is due, per app. Keyed by app so switching into a distraction is
        # checked promptly rather than inheriting the previous app's countdown, while still
        # rate-limiting each app so flicking back and forth can't force a check storm.
        self._next_check_at = {}
        self._listeners = []

    @classmethod
    def shared(cls) -> "MonitorService":
        with cls._lock:
            if cls._instance is None:
                cls._instance = cls()
            return cls._instance

    def subscribe(self, fn):
        self._listeners.append(fn)

    def _set_status(self, text):
        self.status = text
        for fn in list(self._listeners):
            try:
                fn()
            except Exception:
                pass

    # ---- lifecycle ---------------------------------------------------------------------

    def start(self) -> None:
        with self._lock:
            if self._thread is not None and self._thread.is_alive():
                return
            self._stop.clear()
            self.is_running = True
            self._prefs.monitoring_enabled = True
            self._next_check_at.clear()
            mode = "focus" if self._sessions.is_active else "content-rules"
            self._log.add(f"[run] monitoring started ({mode})"
                          + (" [TEST]" if self._prefs.dry_run else ""))
            self._thread = threading.Thread(target=self._loop, name="guardian-monitor", daemon=True)
            self._thread.start()
        self._set_status("Monitoring active")

    def stop(self) -> None:
        with self._lock:
            self._stop.set()
            self._thread = None
            self.is_running = False
            self._prefs.monitoring_enabled = False
            self._log.add("[stop] monitoring stopped")
        self._set_status("Stopped")

    def _loop(self) -> None:
        while not self._stop.is_set():
            try:
                wait = self._tick()
            except Exception as e:
                # A crash in the loop is itself a bypass; log it and keep going.
                self._log.add(f"[warn] monitor tick failed: {e}")
                wait = IDLE_POLL
            self._stop.wait(max(float(wait), 0.5))

    # ---- one tick ----------------------------------------------------------------------

    def _tick(self) -> float:
        """Do at most one unit of work. Returns seconds to wait before the next tick."""
        if BlockController.shared().is_blocking:
            return IDLE_POLL

        # Display asleep or locked: nobody is looking at anything, and capturing would hand us a
        # black frame that reads as "can't see" and alerts.
        if not screen_state.is_visible():
            if self._screen_was_visible:
                self._screen_was_visible = False
                why = screen_state.reason()
                self._log.add(f"[zzz] screen {why} - monitoring paused")
                self._set_status(f"Screen {why}")
            return IDLE_POLL
        if not self._screen_was_visible:
            self._screen_was_visible = True
            self._log.add("[eye] screen back on - monitoring resumed")
            self._set_status("Monitoring active")

        session = self._sessions.current
        if session is not None:
            return self._focus_tick(session)
        if self._prefs.content_rules_enabled:
            return MIN_GAP if self._content_tick() else IDLE_POLL

        self._set_status("No focus session - not watching")
        return IDLE_POLL

    # ---- focus mode --------------------------------------------------------------------

    def _focus_tick(self, session) -> float:
        interval = float(session.interval_seconds)

        if session.is_paused:
            left = max(session.paused_until - time.time(), 0)
            self._set_status(f"Paused {int(left)}s")
            return min(left + 0.5, interval)

        fg = foreground_app.identifier()

        # A foreground window we're not permitted to query (an elevated app while we run
        # unelevated) is a real blind spot. Never a block, but reported after a few readings.
        if fg == foreground_app.UNREADABLE:
            self._unreadable_streak += 1
            if self._unreadable_streak == 5:
                self._log.add("[blind] foreground app can't be identified (elevated process?)")
                if not self._prefs.dry_run and self._prefs.alert_on_unverifiable:
                    self._send_alert(
                        "unreadable-foreground", "[Lockout] Can't identify the app in front",
                        "Guardian could not identify the app in the foreground on this PC "
                        "(it is running with higher privileges than Guardian).", throttle=True)
            return FOCUS_IDLE_POLL
        self._unreadable_streak = 0

        watchlist = session.watchlist(self._prefs.monitored_apps)
        if not fg or fg not in watchlist:
            self._set_status(self._idle_status(session))
            return FOCUS_IDLE_POLL

        if overrides.is_active(fg):
            return FOCUS_IDLE_POLL

        due_at = self._next_check_at.get(fg, 0)
        now = time.time()
        if now < due_at:
            self._set_status(self._idle_status(session))
            return min(due_at - now, FOCUS_IDLE_POLL)

        name = foreground_app.short_name(fg)
        title = foreground_app.title()
        dry = self._prefs.dry_run

        frame = screen_capturer.capture()
        if frame is None:
            self._handle_unverifiable(fg, name, "screen could not be captured", dry)
            self._next_check_at[fg] = time.time() + interval
            return interval
        if frame_quality.is_unreadable(frame):
            self._handle_unverifiable(fg, name, "screen is blank or capture-protected", dry)
            self._next_check_at[fg] = time.time() + interval
            return interval

        self._set_status(f"Checking {name}...")
        cfg = self._prefs.provider_config()
        started = time.time()
        verdict = providers.evaluate(
            cfg, frame, session.task, name, title,
            extra_notes=self._extra_notes(session, fg),
            on_key_worked=self._prefs.promote_api_key)
        ms = int((time.time() - started) * 1000)

        # Transient backend trouble is never evidence about the user.
        if verdict.transient:
            self._log.add(f"[busy] {name} - AI unavailable ({verdict.reason}) - skipped")
            self._set_status("AI unavailable - retrying")
            self._next_check_at[fg] = time.time() + TRANSIENT_RETRY
            return TRANSIENT_RETRY

        self._next_check_at[fg] = time.time() + interval

        if verdict.undetermined:
            self._handle_unverifiable(fg, name, f"AI could not analyse ({verdict.reason})", dry)
            return interval

        if verdict.on_task:
            self._sessions.record_check(off_task=False)
            judgements.record(session, fg, name, title, "on_task", verdict.reason,
                              verdict.confidence, "allowed", cfg.describe())
            self._log.add(f"[ok] {name} on task ({ms}ms, {cfg.model})")
            self._set_status(self._idle_status(session))
            return interval

        self._sessions.record_check(off_task=True)
        action = "logged" if dry else "blocked"
        jid = judgements.record(session, fg, name, title, "off_task", verdict.reason,
                                verdict.confidence, action, cfg.describe())
        if dry:
            self._log.add(f"[WOULD BLOCK] {name} off task ({ms}ms) - {verdict.reason} [TEST]")
            self._set_status(f"WOULD block {name}")
        else:
            self._log.add(f"[BLOCK] {name} OFF TASK ({ms}ms) - {verdict.reason}")
            self._enforce_off_task(session, fg, name, verdict, jid)
        return interval

    def _idle_status(self, session) -> str:
        left = session.remaining_seconds
        if left is None:
            return f"Focus: {session.task[:40]}"
        return f"Focus: {session.task[:32]} ({int(left // 60)}m left)"

    def _extra_notes(self, session, app_id: str) -> str:
        """The user's standing notes plus anything the learner concluded, capped."""
        notes = [self._prefs.focus_notes.strip()]
        if self._prefs.learning_enabled:
            try:
                from ..focus import learned
                notes.append(learned.prompt_suffix(session.task, app_id))
            except Exception:
                pass                    # the learner is experimental; never break a check
        return "\n".join(n for n in notes if n)[:1500]

    def _enforce_off_task(self, session, identifier: str, name: str, verdict, jid: str) -> None:
        """The accountability level is read from the session, not from a global setting that
        could have drifted since it started."""
        self._prefs.last_violation_at = time.time()
        self._set_status(f"BLOCKED {name}")
        BlockController.shared().show(identifier, name, verdict.reason,
                                      session=session, judgement_id=jid)
        notify_local(f"Off task in {name}", verdict.reason or session.task)

        if session.alerts_partner():
            self._send_alert(
                identifier, f"[Lockout] Off task in {name}",
                f"Task: {session.task}\n\n{name} was blocked.\nReason: {verdict.reason}",
                throttle=False)

    # ---- content-rules mode (opt-in) ---------------------------------------------------

    def _content_tick(self) -> bool:
        """Returns True if a monitored app was on screen and was actually checked."""
        fg = foreground_app.identifier()
        if fg == foreground_app.UNREADABLE:
            self._unreadable_streak += 1
            if self._unreadable_streak == 5:
                self._log.add("[blind] foreground app can't be identified (elevated process?)")
            return False
        self._unreadable_streak = 0

        if not fg or fg not in self._prefs.monitored_apps:
            return False
        if overrides.is_active(fg):
            return False

        dry = self._prefs.dry_run
        name = foreground_app.short_name(fg)
        frame = screen_capturer.capture()
        if frame is None:
            self._handle_unverifiable(fg, name, "screen could not be captured", dry)
            return True
        if frame_quality.is_unreadable(frame):
            self._handle_unverifiable(fg, name, "screen is blank or capture-protected", dry)
            return True

        client = OllamaClient(self._prefs)
        cfg = client.config()
        started = time.time()
        verdict = client.evaluate(frame, cfg)
        ms = int((time.time() - started) * 1000)

        if verdict.violation:
            if dry:
                self._log.add(f"[WOULD BLOCK] {name} ({ms}ms) - {verdict.reason} [TEST]")
            else:
                self._log.add(f"[BLOCK] {name} VIOLATION ({ms}ms) - {verdict.reason}")
                self._prefs.last_violation_at = time.time()
                BlockController.shared().show(identifier=fg, app_name=name, reason=verdict.reason)
                notify_local(f"Guardian blocked {name}", verdict.reason)
                self._send_alert(fg, f"[Lockout] Blocked {name}",
                                 f"Content rule triggered in {name}.\n\nReason: {verdict.reason}",
                                 throttle=False)
            return True
        if verdict.transient:
            self._log.add(f"[busy] {name} AI busy - skipped ({ms}ms) - {verdict.reason}")
            return True
        if verdict.undetermined:
            self._handle_unverifiable(fg, name, f"AI could not analyse ({verdict.reason})", dry)
            return True

        self._log.add(f"[ok] {name} cleared ({ms}ms, {cfg.model})")
        return True

    # ---- shared -------------------------------------------------------------------------

    def _handle_unverifiable(self, identifier: str, name: str, why: str, dry: bool) -> None:
        """A screen we cannot read is never a block. Logs, and alerts when configured."""
        if dry:
            self._log.add(f"[blind] {name} CAN'T SEE - {why} - would alert [TEST]")
            self._set_status(f"Can't see {name}")
            return
        self._log.add(f"[blind] {name} CAN'T SEE - {why}")
        self._set_status(f"Can't see {name}")
        session = self._sessions.current
        alerts = self._prefs.alert_on_unverifiable and (session is None or session.alerts_partner())
        if alerts:
            self._send_alert(identifier, f"[Lockout] Couldn't verify {name}",
                             f"Guardian could not see what was on screen in {name} ({why}).",
                             throttle=True)
        if self._prefs.close_unverifiable:
            foreground_app.hide(identifier)

    def _send_alert(self, key: str, title: str, body: str, throttle: bool) -> None:
        if throttle:
            last = self._last_alert_at.get(key)
            if last is not None and time.time() - last < ALERT_THROTTLE:
                return
            self._last_alert_at[key] = time.time()
        full_body = f"{body}\n\n{datetime.now():%Y-%m-%d %H:%M:%S}"
        cfg = pusher.Config(self._prefs.push_enabled, self._prefs.ntfy_server,
                            self._prefs.ntfy_topic)
        err = pusher.send(cfg, title, full_body)
        if err:
            self._log.add(f"[warn] push failed: {err}")
